README.md
# eslint-config-volebo

```yaml
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
```

[![npm version](https://img.shields.io/npm/v/eslint-config-volebo.svg)](https://www.npmjs.com/package/eslint-config-volebo)
[![Build Status](https://app.travis-ci.com/maxkoryukov/eslint-config-volebo.svg?branch=master)](https://app.travis-ci.com/maxkoryukov/eslint-config-volebo)
[![npm downloads](https://img.shields.io/npm/dm/eslint-config-volebo.svg)](https://www.npmjs.com/package/eslint-config-volebo)
[![Known Vulnerabilities](https://snyk.io/test/github/volebo/eslint-config-volebo/badge.svg)](https://snyk.io/test/github/volebo/eslint-config-volebo)


Common style guide for all Volebo.Net projects

## Install

```shell
npm install eslint-config-volebo -S
```

## Usage

Just paste this to the `.eslintrc.js` in the root of your project:

```javascript
exports = module.exports = {
  'extends': [
    'eslint-config-volebo',
  ],
}
```

### Old `eslint` versions

Long time ago `eslint` didn't have such a good support of shared configs, (see
[this issue](https://github.com/eslint/eslint/issues/3458)), but now it does.

If you are still using the old `eslint` this can help:

```javascript
const path = require('path')

exports = module.exports = {
  'extends': [
    path.join(__dirname,
      'node_modules',
      'eslint-config-volebo',
      'index.js'
    )
  ]
}
```

## License

[LICENSE here](LICENSE)


CHANGELOG.md
# CHANGELOG

```
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
```

--------------------------------------------------------------------------------
>
> All notable changes to this project will be documented in this file.
>
> The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
> and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
>
> **This file MUST be filled only by maintainers, using messages from pull**
> **requests.**
>
> new release template
>
> ## [Unreleased] - yyyy-mm-dd
>
> ### Added
> ### Changed
> ### Fixed
> ### Removed
>
--------------------------------------------------------------------------------

## [3.0.0] - 2022-12-19

### Added

- tests with mocha using programmatic `eslint`

### Changed

**Rules**:

- `no-tabs`: off // because we use tabs in the middle of the LOC
- `indent`: error // now it generates ERROR
- `no-var`: error // yep, no more vars
- `no-const-assign`: error // why not?
- `no-console`: error  // now it is an error. To automatically decline
  contributions (on Travis level) with `console.log` in the PR
- `no-alert`: error // just the same as `no-console`
- `no-param-reassign`: error  // preventing side effects

**ALSO (non rules)**:

- add `parserOptions.ecmaVersion` (by default we use js.es9 == es2018)

--------------------------------------------------------------------------------

## [2.3.5] - 2020-07-13
## [2.3.4] - 2020-07-13

### Added

- `semi` - don't use `;` (see volebo/eslint-config-volebo#7)
- `quotes` - singlequotes (fix volebo/eslint-config-volebo#5)
- `no-console` - why do you want to print to console?
- `no-plusplus` - be careful with this

### Changed

- `no-unused-vars` - now unused function's arguments could start with `__xxx` 

--------------------------------------------------------------------------------

## [2.3.3] - 2016-06-15

--------------------------------------------------------------------------------

## [0.0.1] - 2016-06-15

INIT.


test.sh
#!/usr/bin/env bash

################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################

ESLINT="eslint -c index.js"

$ESLINT test/example.01.good.js

exit 0


tests/test.sh
#!/usr/bin/env bash

################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
#
# Copyright (C) 2016-2020 Volebo <dev@volebo.net>
# Copyright (C) 2016-2020 Maksim Koryukov <maxkoryukov@volebo.net>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the MIT License, attached to this software package.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# You should have received a copy of the MIT License along with this
# program. If not, see <https://opensource.org/licenses/MIT>.
#
# http://spdx.org/licenses/MIT
#
################################################################################

ESLINT="`npm bin`/eslint -c index.js"

$ESLINT tests/example.01.good.js

# ==============================================================================

# $ESLINT tests/example.02.bad.js > $

exit 0


tests/bootstrap.js
/*
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
#
# DESCRIPTION
#
# Copyright (C) 2016-2022 Volebo <dev@volebo.net>
# Copyright (C) 2016-2022 Maksim Koryukov <maxkoryukov@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the MIT License, attached to this software package.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# You should have received a copy of the MIT License along with this
# program. If not, see <https://opensource.org/licenses/MIT>.
#
# http://spdx.org/licenses/MIT
#
################################################################################
*/

'use strict'

const chai                         = require('chai')
// const path                         = require('path')
// const mh                           = require('@volebo/mocha-helpers')


global.expect = chai.expect

// global.filename2suitename = mh.filename2suitename
// global.tags = mh.tags


// const report_path = path.join(
// 	process.cwd(),
// 	'tmp',
// 	'mochawesome-report'
// )

// process.env.MOCHAWESOME_REPORTDIR = report_path


.mocharc.js
/*
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
#
# Copyright (C) 2016-2022 Volebo <dev@volebo.net>
# Copyright (C) 2016-2022 Maksim Koryukov <maxkoryukov@volebo.net>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the MIT License, attached to this software package.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# You should have received a copy of the MIT License along with this
# program. If not, see <https://opensource.org/licenses/MIT>.
#
# http://spdx.org/licenses/MIT
#
################################################################################
*/


const MOCHA_TIMEOUT_DEFAULT = Number(process.env.MOCHA_TIMEOUT_DEFAULT || 2000)


module.exports = {
	reporter: 'list',
	// reporter: 'spec',

	recursive: true,
	checkLeaks: true,
	ui: 'bdd',
	require: 'tests/bootstrap',
	extension: [
		'test.js',
		'spec.js',
	],
	retries: 1,

	diff: true,
	//opts: './test/mocha.opts',
	package: './package.json',
	slow: 75,
	timeout: MOCHA_TIMEOUT_DEFAULT,

	spec: [
		'tests/**/*.test.js',
		// 'src/**/*.test.js',
		// 'src/**/*.spec.js',

		// 'libs/**/*.test.js',
	],
}


.eslintrc.js
/*
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
#
# Copyright (C) 2016-2022 Volebo <dev@volebo.net>
# Copyright (C) 2016-2022 Maksim Koryukov <maxkoryukov@volebo.net>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the MIT License, attached to this software package.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# You should have received a copy of the MIT License along with this
# program. If not, see <https://opensource.org/licenses/MIT>.
#
# http://spdx.org/licenses/MIT
#
################################################################################
*/

'use strict'


const path = require('path')

exports = module.exports = {
	'root': true,

	'env': {
		node: true
	},

	'parserOptions': {
		// 'parser': 'babel-eslint'
		'ecmaVersion': 2018,
	},

	'extends': [
		// 'plugin:vue/essential',
		// 'volebo',
		path.join(__dirname, 'index.js'),
	],

	'overrides': [
		{
			'files': [
				'**/*.test.js',
				'**/*.spec.js'
			],

			'env': {
				mocha: true
			},

			'globals': {
				'filename2suitename': true,
				'tags': true,
			},

		}
	],
}


index.js
/*
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
#
# Copyright (C) 2016-2022 Volebo <dev@volebo.net>
# Copyright (C) 2016-2022 Maksim Koryukov <maxkoryukov@volebo.net>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the MIT License, attached to this software package.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
# You should have received a copy of the MIT License along with this
# program. If not, see <https://opensource.org/licenses/MIT>.
#
# http://spdx.org/licenses/MIT
#
################################################################################
*/


'use strict'


exports = module.exports = {
	extends: [
		'eslint:recommended'
		//'./node_modules/coding-standard/eslintDefaults.js',
		// Override eslintDefaults.js
		//'./node_modules/coding-standard/.eslintrc-es6',
		// Override .eslintrc-es6
		//'./node_modules/coding-standard/.eslintrc-jsx',
	],

	env: {
		'node': true,
		'es6': true,
	},

	parserOptions: {
		// https://kangax.github.io/compat-table/es2016plus/https://kangax.github.io/compat-table/es2016plus/
		ecmaVersion: 9,

		// sourceType: 'module',
	},

	rules: {
		// Override any settings from the 'parent' configuration
		'comma-dangle' : ['error', 'only-multiline'],  // TODO: review this rule
		'no-process-exit': ['error'],

		/* CODE STYLE */
		'dot-notation': 'off',

		'indent': [
			'error',
			'tab',
			{
				'SwitchCase': 1,
			}
		],
		'new-cap': ['warn'],
		'no-extra-parens': ['warn', 'all', { 'nestedBinaryExpressions': false }],
		'curly': 'error',
		'no-trailing-spaces': 'error',
		'no-irregular-whitespace': ['error', { 'skipComments': true }],
		'no-new-object': 'error',
		'semi': [ 'error', 'never'],
		'quotes': [
			'error',
			'single',
			{
				avoidEscape: true,
				allowTemplateLiterals: true,
			},
		],
		'no-tabs': 'off',

		/* STATIC SECURITY */
		'strict': ['error', 'global'],

		'no-cond-assign': 'error',
		'eqeqeq': 'error',
		'no-bitwise': 'error',
		'guard-for-in': 'error',
		'no-eq-null': 'error',
		'no-unreachable': 'error',
		'no-eval': 'error',
		'no-loop-func': 'error',
		'no-iterator': 'error',
		'no-multi-str': 'error', // use \n instead!
		'valid-typeof': 'error',
		'no-proto': 'error',
		'no-script-url': 'error',
		'require-yield': 'error',

		'no-native-reassign': 'error',
		'no-extend-native': 'error',

		'max-params': ['warn', { 'max': 4 }],
		'complexity': ['warn', { 'max': 20 }],
		'max-depth': ['warn', { 'max': 5 }],

		'no-path-concat': ['error'],

		'no-caller': ['error'],
		'no-empty': ['error', { 'allowEmptyCatch': false }],
		'no-empty-function': ['error'],
		'no-new': ['error'],

		// variables definitions
		'no-var': 'error',
		'prefer-const': ['warn', { /*'ignoreReadBeforeAssign' : false */ }],
		'no-const-assign': 'error',
		'no-use-before-define': ['error', { 'functions': false, 'classes': true }],
		'no-undef': ['error', { 'typeof': true }],

		// will ignore unused arguments, but skip args:
		// `_` - lodash placeholder
		// `next` - express callback
		// `_unused_XXX` - explicitly unused named variables
		'no-unused-vars' : ['error', { 'args': 'all', 'argsIgnorePattern': '(_|next|^__)' }],

		/* SMELL */
		'no-console': 'error',  // use `debug` or other packages, or add explicit "eslint-ignore" for lines with console.xxx
		'no-alert': 'error',

		'no-plusplus': [
			'error',
			{
				allowForLoopAfterthoughts: true,
			},
		],
		'no-param-reassign': [
			'error',
			{
				props: false,
			},
		],

	},

	overrides: [
		{
			files: [
				'**/*.test.js',
				'**/*.spec.js'
			],

			env: {
				'mocha': true,
			},
			globals: {
				'expect': true,
				'filename2suitename': true,
				'tags': true,
			},
			rules: {
				// `expectJS` uses expressions like `.xx.is.null`
				'no-unused-expressions': 'off',
			},
		}
	],
}
